### ISSUE TITLE
[Bug] `datetime.datetime.utcnow()` is deprecated as of Python 3.12

### ISSUE THREAD

**Issue by dataders** [dbt-labs/dbt-core#9791]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Thanks to [Talla in #adapter-ecosytem Slack channel](https://getdbt.slack.com/archives/C030A0UF5LM/p1711010240341219) for the heads up

Any usage of `datetime.datetime.utcnow()`  throws the following deprecation warning in Python 3.12

> DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).

related: https://github.com/dbt-labs/dbt-common/issues/99

### Expected Behavior

This shouldn't happen

### Steps To Reproduce

Run our unit test suite with Python 3.12

### Environment

```markdown
- OS:
- Python: 3.12
- dbt: `1.7.10`
```

**Comment by dbeatty10** [comment]
You are the king @dataders 👑

Potential solution here: https://github.com/dbt-labs/dbt-core/issues/5267#issuecomment-1513306428

**Comment by dbeatty10** [comment]
### TLDR

It looks to me like we can preserve the current behavior by updating references to:
```python
datetime.utcnow()
```
with
```python
datetime.now(timezone.utc).replace(tzinfo=None)
```

### Recommendations

Short term, I'd recommend:
- Resolve deprecation warnings and preserve current behavior (using naive datetimes everywhere)

Long-term, I'd recommend:
- Use aware datetimes everywhere (choosing appropriate string formatting upon rendering) as described in [#5267](https://github.com/dbt-labs/dbt-core/issues/5267#issuecomment-1513306428)

### Occurrences

This shows up in our code base in a lot of places:
- https://github.com/search?q=repo%3Adbt-labs%2Fdbt-core+%22utcnow%28%29%22&type=code&p=1
  - See also: https://github.com/search?q=repo%3Adbt-labs%2Fdbt-core+utcnow&type=code

### More detail

```python
# compare_datetimes.py

from datetime import timezone, datetime


# To compare/contrast effectively, set your local system time zone to something _different_ than UTC 

# Naive datetimes -- beware!
naive_1 = datetime.utcnow()  # ❌ current
naive_2 = datetime.now(timezone.utc).replace(tzinfo=None)  # ✅ recommended search/replace
naive_3 = datetime.now()  # extra beware this one!

# Aware datetimes
aware_1 = datetime.now(timezone.utc)  # standardized to UTC
aware_2 = datetime.now().astimezone()  # system local time zone

naive = [naive_1, naive_2, naive_3]
aware = [aware_1, aware_2]

for i in naive + aware:
    # Format the datetime as a string using the ISO 8601 format
    print(i.isoformat())
```

Run the python script:
```shell
python compare_datetimes.py
```

Output (using [`America/Boise`](https://en.wikipedia.org/wiki/America/Boise) from [tz database](https://en.wikipedia.org/wiki/Tz_database)):
```
2024-03-22T16:07:07.997093
2024-03-22T16:07:07.997582
2024-03-22T10:07:07.997598
2024-03-22T16:07:07.997601+00:00
2024-03-22T10:07:07.997602-06:00
```

**Comment by slothkong** [comment]
Hello there,

I found myself with some spare time during the Easter break and decided to follow through on this issue. Please forgive any impudence🙏

@dbeatty10, I second your thoughts. Found very sensible [your comment](https://github.com/dbt-labs/dbt-core/issues/9791#issuecomment-2015447724), about decoupling `datetime.utcnow()` deprecation from feature requests with somewhat unclear scope and action-plan. 
Furthermore, I went over your `compare_datetimes.py` script, as well as the comments you left in #5267 and the [official datetime docs](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow). All signs indeed point to `datetime.now(timezone.utc).replace(tzinfo=None)`  as the recommended replacement, mainly because it safeguards cases in which `dbt` executes in environments where wrong assumptions about system time setup might lead to unexpected behavior.

@dataders, [to your helpful diagnose](https://github.com/dbt-labs/dbt-core/issues/9791#issue-2200840508), I just wanted to add that besides `dbt-common`, `logback` is the other package causing the deprecation warnings. In fact, looks like the devs over there had dealt with this topic as of release 1.7.0, by wrapping the recommended replacement statement into a [function](https://github.com/getlogbook/logbook/blob/e7cc074e6a7affa86c033efb08591e59356d65b9/src/logbook/helpers.py#L147-L157):

```python
if sys.version_info >= (3, 12):

    def datetime_utcnow():
        return datetime.now(timezone.utc).replace(tzinfo=None)

else:
    datetime_utcnow = datetime.utcnow
```


thus saving some characters in the places [where UTC time is needed](https://github.com/getlogbook/logbook/blob/e7cc074e6a7affa86c033efb08591e59356d65b9/src/logbook/handlers.py#L509), like this:
```python
first_count = now = datetime_utcnow()
```

In a future where `dbt` has standalone configuration for users to define the timezone they want for their logs, maybe a simple wrapper like that could help wire the user input across the code base. 

But I digress... 

To get some results firsts, I just want to help on:
1. the execution of your recommendations @dbeatty10 
2. bringing a bit of homogeneity to the ways  `datetime` and its classes are being imported, such that future development would require less effort

With those two goals in mind, I put up [fix branch](https://github.com/dbt-labs/dbt-core/compare/main...slothkong:dbt-core:fix_9791) on my fork, all code changes included.
Both unit and integration tests pass. I also created a local HTML coverage report before and after the changes, and saw no coverage degradation whatsoever, so I think codecov won't complain 🤞

The deprecation warnings are gone, except for:
* Those of `dbt-common` which probably be fixed [issue 99](https://github.com/dbt-labs/dbt-common/issues/99)
* Those of `logback` 

To address the latter, I tried bumping the package version from:
https://github.com/dbt-labs/dbt-core/blob/71f3519611a6a21c2cb492df2d54e60af040e0f3/core/setup.py#L58
to 
```python
"logbook>=1.5,<1.7.0.post0",
```
However, for reasons I still do not understand, `tox` does not seem to realize I want it to use version 1.7.0 at test time. This is actually the only reason why I did not open a pull request yet. Any suggestions or comments are more than welcomed.

Cheers~

**Comment by slothkong** [comment]
Upon closer look, the issue was not that `tox` refused to use the correct version of `logbook`, rather that even the latest version of the package was using another related method, also deprecated (`datetime.utcfromtimestamp()`). 

The folk over there have been informed. Until a new release is done on their end, there is not much that can be done for this issue, so I will keep the versions that is pinned currently (`logbook>=1.5,<1.6`) and open a pull request with code changes specific only to `dbt-core`  🙈

**Comment by KanbanGroup** [comment]
This issue is a serious pest when you are running in the foreground .. i.e. developer mode without supervisor. It wont affect regular users, but as a developer, when I am looking for my print() statements to display debugging info, it is a real pest to have it all scrolled off the screen with a whole load of unnecessary deprecation warnings about "datetime.datetime.utcnow".

Since it will have to be fixed once the method is dropped, would it not be better to fix it while there is time to do the work?
I would fix it myself and offer up the solution for merging, but I just don't trust my git skills that far, and I am only working on custom mods for the company where I use frappe/erpnext as a business tool.

**Comment by sdebruyn** [comment]
The thing is deprecated and will be removed in a future Python version. It seems very strange to label this as "not planned".

**Comment by slothkong** [comment]
@sdebruyn 
The core team has been working on refactoring, some of which partly overlaps with this issue - the must obvious example I could find being removal of the legacy logger https://github.com/dbt-labs/dbt-core/pull/10119 .

Without having any more insights than you, my gut feeling tells me this issue might be tackled as part of some of the refactoring epic, such as https://github.com/dbt-labs/dbt-core/issues/11335 , later on.

**Comment by sdebruyn** [comment]
Given that this is an OSS / community project, the transparancy shouldb be better. IMHO closing an issue means you're not going to fix it. In this case, it should have stayed open and should have been linked to #10119 and #11335. The stalebot comes accross as evil. #11335 would definitely be necessary as even with #10119 merged in, I still get 10K+ warnings when I run the integration tests in my adapter.

**Comment by slothkong** [comment]
@sdebruyn, from my side all I can say is that, if the issue gets reopened, I am willing to rework https://github.com/dbt-labs/dbt-core/pull/9839 to resolve the merge conflicts that pilled up since it was set to "Not Planned", and hope for the best during the code review 🤞
